lib/truffle/java/java_ext/java_util_list.rb

```diff
@@ -0,0 +1,153 @@
+# Copyright (c) 2017 Oracle and/or its affiliates. All rights reserved. This
+# code is released under a tri EPL/GPL/LGPL license. You can use it,
+# redistribute it and/or modify it under the terms of the:
+#
+# Eclipse Public License version 1.0
+# GNU General Public License version 2
+# GNU Lesser General Public License version 2.1
+
+java.util.List
+
+module ::Java::JavaUtil::List
+  def [](start, length=nil)
+    size = self.size
+    if start.kind_of?(Range)
+      first = start.first
+      last = start.last
+      last += size if last < 0
+      first ++ size if first < 0
+      return nil if first < 0 || first >= size
+      last += 1 unless start.exclude_end?
+      sub_list(first, last)
+    else
+      start += size if start < 0
+      return nil if start < 0 || start >= size
+      if length == nil
+        get(start)
+      else
+        return nil if length < 0
+        last = length + start
+        last = size if last > size
+        sub_list(start, last)
+      end
+    end
+  end
+
+  def []=(index, value)
+    size = self.size
+    if index.kind_of?(Range)
+      first = index.first
+      last = index.last
+      first += size if first < 0
+      last += size if last < 0
+      last += 1 unless index.exclude_end?
```

Comment:
That would be nice but the handling of negative end values is in the array code, and it is split between `Array#[]` in Ruby and 'ArrayNode.IndexNode.slice' in Java.